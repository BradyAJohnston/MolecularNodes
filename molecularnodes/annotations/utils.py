import inspect
import types
import typing
import bpy
from PIL import Image


def get_all_class_annotations(cls) -> dict:
    """Get all the annotations from the class including base classes"""
    all_annotations = {}
    for base_cls in cls.mro():
        if hasattr(base_cls, "__annotations__"):
            current_annotations = inspect.get_annotations(base_cls, eval_str=True)
            all_annotations.update(current_annotations)
    if "interface" in all_annotations:
        del all_annotations["interface"]
    return all_annotations


def get_blender_supported_type(atype):
    supported_types = (
        str,
        bool,
        int,
        float,
        tuple[float, float],
        tuple[float, float, float],
        tuple[float, float, float, float],
    )
    if atype in supported_types:
        return atype
    if typing.get_origin(atype) is typing.Union or type(atype) is types.UnionType:
        for utype in atype.__args__:
            if utype in supported_types:
                return utype
    return None


def get_view_matrix(obj):
    if obj._render_mode:
        return obj._scene.camera.matrix_world.inverted()
    else:
        return obj._rv3d.view_matrix


def is_perspective_projection(obj):
    if obj._render_mode:
        return obj._scene.camera.data.type != "ORTHO"
    else:
        return obj._rv3d.is_perspective


def render_annotations(
    scene: bpy.types.Scene, render_scale: float, image: Image, image_scale: float
) -> None:
    """Render annotations of all entities to an image"""
    session = scene.MNSession
    session.prune()  # remove any invalid session entities
    for entity in session.entities.values():
        if not hasattr(entity, "annotations"):
            continue
        manager = entity.annotations
        manager._enable_render_mode(scene, render_scale, image, image_scale)
        manager._draw_annotations()
        manager._disable_render_mode()
