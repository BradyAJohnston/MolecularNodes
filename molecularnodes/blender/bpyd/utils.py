import numpy as np


def centre(position: np.ndarray, weight: np.ndarray | None = None):
    "Calculate the weighted centroid of the vectors"
    if weight is None:
        return np.mean(position, axis=0)
    return np.sum(position * weight.reshape((-1, 1)), axis=0) / np.sum(weight)
