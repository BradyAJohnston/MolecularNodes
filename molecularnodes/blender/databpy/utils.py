import bpy


def evaluate_object(obj: bpy.types.Object):
    "Return an object which has the modifiers evaluated."
    obj.update_tag()
    return obj.evaluated_get(bpy.context.evaluated_depsgraph_get())
