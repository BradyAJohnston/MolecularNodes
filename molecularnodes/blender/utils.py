import bpy
from pathlib import Path


def path_resolve(path: str | Path) -> Path:
    if isinstance(path, str):
        return Path(bpy.path.abspath(path))
    elif isinstance(path, Path):
        return Path(bpy.path.abspath(str(path)))
    else:
        raise ValueError(f"Unable to resolve path: {path}")
