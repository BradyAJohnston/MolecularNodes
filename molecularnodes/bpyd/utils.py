import numpy as np


def centre(position: np.ndarray, weight: np.ndarray | None = None):
    "Calculate the weighted centroid of the vectors"
    if weight is None:
        return np.mean(position, axis=0)
    return np.sum(position * weight.reshape((-1, 1)), axis=0) / np.sum(weight)


def lerp(a: np.ndarray, b: np.ndarray, t: float = 0.5) -> np.ndarray:
    """
    Linearly interpolate between two values.

    Parameters
    ----------
    a : array_like
        The starting value.
    b : array_like
        The ending value.
    t : float, optional
        The interpolation parameter. Default is 0.5.

    Returns
    -------
    array_like
        The interpolated value(s).

    Notes
    -----
    This function performs linear interpolation between `a` and `b` using the
    interpolation parameter `t` such that the result lies between `a` and `b`.

    Examples
    --------
    >>> lerp(1, 2, 0.5)
    1.5

    >>> lerp(3, 7, 0.2)
    3.8

    >>> lerp([1, 2, 3], [4, 5, 6], 0.5)
    array([2.5, 3.5, 4.5])

    """
    return np.add(a, np.multiply(np.subtract(b, a), t))
