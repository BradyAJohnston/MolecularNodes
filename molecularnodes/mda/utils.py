"""
Utils
"""

import bpy
from bpy.app.handlers import persistent  # type: ignore


def is_mda_universe(object: bpy.types.Object) -> bool:
    """Check if blender object is MDA universe"""
    return object and object.mn.mda and object.mn.mda.is_mda_universe


def _msgbus_active_object_callback(*args):
    """Callback when active object changes"""
    sprop = bpy.context.scene.mn.mda
    active_object = bpy.context.active_object
    index = -1
    if is_mda_universe(active_object):
        # find index of universe in scene universes collection
        index = sprop.universes.find(active_object.mn.mda.universe_key)
    sprop.active_index = index  # updates selection in Universes panel


def _subscribe_to_active_object():
    """Actual subscribe to active object changes"""
    subscribe_to = (bpy.types.LayerObjects, "active")
    bpy.msgbus.subscribe_rna(
        key=subscribe_to,
        owner=bpy,
        args=(),
        notify=_msgbus_active_object_callback,
    )


@persistent
def active_object_load_handler(dummy):
    """Load handler to re-subscribe to active object changes"""
    _subscribe_to_active_object()


def subscribe_to_active_object():
    """Subscribe to active object changes"""
    _subscribe_to_active_object()  # subscribe
    # create a load handler to re-subscribe on file loads
    bpy.app.handlers.load_post.append(active_object_load_handler)


def set_object_visibility(object: bpy.types.Object, visible: bool) -> None:
    """Set visibility of Blender object"""
    if object.name not in bpy.context.view_layer.objects:
        return
    try:
        object.hide_set(not visible)
        # obj.hide_viewport = hide
        # icon incorrect, but causes blender bug for volumes with above
        object.hide_render = not visible
    except:
        # Keyframing object visibility can at times lead to:
        # RuntimeError: Object can't be hidden because it is not in View Layer 'ViewLayer'!
        pass
