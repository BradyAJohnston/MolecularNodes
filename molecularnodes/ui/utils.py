import bpy

def check_online_access_for_ui(layout: bpy.types.UILayout) -> bpy.types.UILayout:
    """
    Disable UI without Online Access

    Checks for the online access permissions, and adds a warning and disables following
    UI elements if it fails the check. Returns the UILayout that will have .enabled flag
    set to False, disabling all subsequent uses of the layout.

    Args:
        layout (bpy.types.UILayout): The UILayout element to add the warning and potentially
        disable.

    Returns:
        bpy.types.UILayout: The altered UILayout element, for use in downstream UI
        components.
    """
    if not bpy.app.online_access:
        layout.label(
            text="Online access disabled. Change in Blender's system preferences.",
            icon="ERROR",
        )
        op = layout.operator("wm.url_open", text="Online Access Docs", icon="URL")
        op.url = "https://docs.blender.org/manual/en/dev/editors/preferences/system.html#bpy-types-preferencessystem-use-online-access"
        layout = layout.column()
        layout.alert = True
        layout.enabled = False

    return layout