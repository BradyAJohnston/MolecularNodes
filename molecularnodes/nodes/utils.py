import bpy
import databpy


def get_links(socket: bpy.types.NodeSocket) -> bpy.types.NodeLinks:
    """
    Get the links between two sockets
    """
    links = socket.links
    if links is None:
        raise ValueError("No links found for socket")
    return links


def int_menu_switch_from_dict(
    dictionary: dict, name: str = "NewMenuSwitch"
) -> bpy.types.NodeTree:
    """
    Create a menu from a dictionary of options
    """
    tree = databpy.nodes.new_tree(name=name, geometry=False)

    node_m = tree.nodes.new("GeometryNodeMenuSwitch")
    node_m.data_type = "INT"

    starting_enum_length = len(node_m.enum_items)

    for i, (key, value) in enumerate(dictionary.items()):
        if i < starting_enum_length:
            node_m.enum_items[i].name = key
        else:
            node_m.enum_items.new(name=key)

        node_m.inputs[i + 1].default_value = value

    tree.interface.new_socket("Selector", socket_type="NodeSocketMenu")
    tree.interface.new_socket("Value", socket_type="NodeSocketInt", in_out="OUTPUT")
    tree.links.new(databpy.nodes.get_input(tree).outputs["Selector"], node_m.inputs[0])
    tree.links.new(
        node_m.outputs["Output"], databpy.nodes.get_output(tree).inputs["Value"]
    )
    return tree
