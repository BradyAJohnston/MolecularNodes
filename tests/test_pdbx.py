import molecularnodes as mn
import numpy as np

import random
from .constants import data_dir
from .utils import NumpySnapshotExtension, sample_attribute


def test_ss_label_to_int():
    examples = ['TURN_TY1_P68', 'BEND64', 'HELX_LH_PP_P9', 'STRN44']
    assert [3, 3, 1, 2] == [
        mn.io.parse.cif._ss_label_to_int(x) for x in examples]


def test_get_ss_from_mmcif(snapshot_custom: NumpySnapshotExtension):
    mol = mn.io.load(data_dir / '1cd3.cif')

    # mol2, fil2 = mn.io.fetch('1cd3')

    random.seed(6)
    random_idx = random.sample(range(len(mol)), 100)

    # assert (mol.sec_struct == mol2.sec_struct)[random_idx].all()

    assert snapshot_custom == mol.array.sec_struct[random_idx]


def test_secondary_structure_no_helix(snapshot_custom):
    m = mn.io.fetch('7ZL4', cache_dir=data_dir)

    assert snapshot_custom == sample_attribute(
        m.object, 'sec_struct', n=500, evaluate=False)
