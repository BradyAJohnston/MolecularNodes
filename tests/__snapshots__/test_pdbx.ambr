# serializer version: 1
# name: test_get_ss_from_mmcif
  [3 1 3 1 1 3 1 3 2 3 3 1 3 3 3 3 2 3 1 1 3 2 1 3 1 3 1 3 2 3 3 1 3 2 1 3 3
   3 3 2 1 2 3 2 1 3 3 1 1 3 3 3 1 3 3 3 3 1 1 3 3 3 1 1 3 2 1 1 1 3 1 3 3 1
   1 3 3 2 1 3 1 3 3 3 3 1 3 3 1 3 3 3 1 3 3 2 1 3 3 1]
# ---
# name: test_secondary_structure_no_helix
  [2 2 3 3 3 2 2 3 3 2 3 2 3 2 3 3 2 2 2 3 2 2 3 2 2 2 2 2 2 2 2 3 2 3 2 2 2
   3 2 2 3 2 2 2 2 2 2 2 2 3 2 3 3 3 2 2 2 2 3 3 2 3 2 3 3 2 2 3 3 2 2 2 2 2
   3 3 3 2 3 3 2 2 2 3 2 2 2 3 2 3 2 3 3 3 2 2 3 3 3 2 2 2 2 3 2 3 2 2 3 2 2
   2 3 3 2 2 2 3 3 3 3 3 3 3 3 3 2 3 3 2 2 2 3 2 2 2 3 2 2 3 3 2 2 3 3 3 2 3
   3 2 2 3 2 2 2 2 2 3 2 3 2 2 3 2 2 2 3 3 2 2 3 3 3 3 3 3 3 3 2 2 3 2 2 2 3
   3 2 2 3 2 3 3 2 3 3 3 3 3 3 2 3 2 3 2 3 3 2 3 2 2 3 2 2 2 2 3 2 2 3 3 3 3
   3 2 2 2 3 2 3 2 3 2 2 3 2 3 2 2 3 3 2 3 3 2 3 3 3 3 3 2 2 3 3 2 3 3 3 2 3
   3 2 2 2 3 2 2 3 2 2 2 3 3 3 2 3 2 3 2 3 3 2 2 2 2 2 2 2 3 2 2 3 2 2 3 2 2
   2 3 3 2 2 2 2 2 2 2 3 2 3 2 3 3 2 3 3 3 2 2 3 3 2 3 3 2 3 2 2 2 2 2 3 2 2
   2 2 3 2 3 3 3 2 3 3 2 2 2 2 2 2 3 2 2 3 3 3 3 3 3 2 2 2 2 2 2 2 3 3 3 3 2
   3 2 3 2 3 2 3 2 2 3 3 2 2 3 3 3 3 3 3 3 3 2 3 2 2 2 2 2 3 3 2 3 3 2 2 2 3
   2 2 3 2 3 2 3 2 3 2 2 2 2 2 2 3 3 2 3 2 3 3 2 3 2 2 3 2 3 3 3 3 3 2 2 2 3
   2 2 3 3 3 2 3 2 2 2 3 3 2 3 3 2 3 2 3 3 3 2 2 3 2 2 3 2 3 3 3 3 2 2 2 3 3
   3 2 2 2 2 2 3 2 2 2 2 3 3 2 3 2 3 3 3]
# ---
